vault/seal/ocikms/ocikms.go

```diff
@@ -0,0 +1,305 @@
+// Copyright © 2019, Oracle and/or its affiliates.
+package ocikms
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"fmt"
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/errwrap"
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/vault/sdk/physical"
+	"github.com/hashicorp/vault/vault/seal"
+	"github.com/oracle/oci-go-sdk/common"
+	"github.com/oracle/oci-go-sdk/common/auth"
+	"github.com/oracle/oci-go-sdk/keymanagement"
+	"math"
+	"os"
+	"strconv"
+	"time"
+)
+
+const (
+	// OCI KMS key ID to use for encryption and decryption
+	EnvOCIKMSSealKeyID = "VAULT_OCIKMS_SEAL_KEY_ID"
+	// OCI KMS key ID to use for encryption and decryption
+	EnvOCIKMSCryptoEndpoint = "VAULT_OCIKMS_CRYPTO_ENDPOINT"
+	// Should retry for calls to KMS endpoints
+	KMSShouldRetryOperation = true
+	// Maximum number of retries
+	KMSMaximumNumberOfRetries = 5
+	// keyID config
+	KMSConfigKeyID = "keyID"
+	// cryptoEndpoint config
+	KMSConfigCryptoEndpoint = "cryptoEndpoint"
+	// authTypeAPIKey config
+	KMSConfigAuthTypeAPIKey = "authTypeAPIKey"
+)
+
+var (
+	metricInit    = []string{"ocikms", "init"}
+	metricEncrypt = []string{"ocikms", "encrypt"}
+	metricDecrypt = []string{"ocikms", "decrypt"}
+
+	metricInitFailed    = []string{"ocikms", "initFailed"}
+	metricEncryptFailed = []string{"ocikms", "encryptFailed"}
+	metricDecryptFailed = []string{"ocikms", "decryptFailed"}
+)
+
+// OCIKMSMechanism is the method used to encrypt/decrypt in auto unseal process
+type OCIKMSMechanism uint32
+
+type OCIKMSSeal struct {
+	authTypeAPIKey bool                           // true for user principal, false for instance principal, default is false
+	keyID          string                         // OCI KMS keyID
+	cryptoEndpoint string                         // OCI KMS crypto endpoint
+	cryptoClient   *keymanagement.KmsCryptoClient // OCI KMS crypto client
+	logger         log.Logger
+}
+
+var _ seal.Access = (*OCIKMSSeal)(nil)
+
+// NewSeal creates a new OCIKMSSeal seal with the provided logger
+func NewSeal(logger log.Logger) *OCIKMSSeal {
+	k := &OCIKMSSeal{
+		logger: logger,
+		keyID:  "",
+	}
+	return k
+}
+
+func (k *OCIKMSSeal) SetConfig(config map[string]string) (map[string]string, error) {
+	defer metrics.MeasureSince(metricInit, time.Now())
+	if config == nil {
+		config = map[string]string{}
+	}
+
+	// Check and set KeyID
+	switch {
+	case os.Getenv(EnvOCIKMSSealKeyID) != "":
+		k.keyID = os.Getenv(EnvOCIKMSSealKeyID)
+	case config[KMSConfigKeyID] != "":
+		k.keyID = config[KMSConfigKeyID]
+	default:
+		metrics.SetGauge(metricInitFailed, 1)
+		return nil, fmt.Errorf("'%s' not found for OCI KMS seal configuration", KMSConfigKeyID)
+	}
+	k.logger.Info(fmt.Sprintf("got %s: %s", KMSConfigKeyID, k.keyID))
```

Comment:
I don't think we want the KMS details showing up in logs. If we do keep this, it would be cleaner like:
`k.logger.Info("OCI kms configuration", KMSConfigKeyID, k.keyID)`